### ISSUE TITLE
favorites and downloaded chapters wiped

### ISSUE THREAD

**Issue by nigredo-tori** [riderkick/FMD#218]
This morning I have opened FMD just to find my favorites and finished downloads list empty. It might have been an update (I honestly don't remember whether I installed one yesterday). The things I managed to find out:
1) My downloadedchapters.ini is now two hundred kilobytes of 0x00 character, followed by 0D 0A.
2) I have 0 favorites.
3) My works.ini is empty (I have no idea if there was anything there before).
4) My websites tab seems to have reverted to defaults
5) Few other settings I happen to remember appear to be untouched (download folders, Bato.to account etc.).

**Comment by nigredo-tori** [comment]
I vaguely remember killing FMD after it hanged/deadlocked (it does this sometimes), so this is probably for the same reason as #153

**Comment by riderkick** [comment]
How to reproduce the problem?
I can't reasoning how downloadchapters.ini, favorites.ini and works.ini broken at the same time. They are exist in main thread, and works one at a time.

**Comment by nigredo-tori** [comment]
I can't reproduce the problem right now - can't make FMD hang long enough on exit... Reproducing would probably involve large number of downloaded chapters, favorites, and some kind of IO load such that FMD would appear unresponsive.
As to why all these files are corrupted at once - it may have something to do with the fact I restarted FMD after the bug (the old turning off and on routine). Seeing as it stores its state on exit regardless of whether it changed, I suppose FMD could read some kind of garbage from one of the files, and propagate it somehow?
It seems to me like the root of the problem is the window of time when state files are inconsistent (while they are written). If something happens during that window (a process kill, or some kind of crash), the state would be corrupted with no way to restore it. This window gets larger with more state to store, so the probability of something like this happening increases.
A simple solution would be writing the state in a new file (e.g. downloadedchapters.ini.1), and replacing the old one once finished. This way if the program dies while saving state, no corruption should occur (barring the case when we kill the program between deleting the old file and moving the new).

**Comment by riderkick** [comment]
It was TIniFile, which work on memory and only written to disk when updatefile called. updatefile only called when exit and ontimer, which is all work in main thread. so there is only one executed at a time.
When app hang=main thread hang. So there is only one fault at a time. can't break all at once.

**Comment by riderkick** [comment]
Now config, works and favorites using .run file. I don't know if that will fix the issue.
downloadedchapters now using sqlite3 database, so there is shouldn't be any issue.
Please check another issue too. #201, #204

**Comment by alvarexoz** [comment]
well, i still have those problem.
![crash on mangapark](https://cloud.githubusercontent.com/assets/17180171/14228184/56934102-f938-11e5-8b65-2125bdfd50c9.png)

​clearly, no checkbox on it. [#204](https://github.com/riderkick/FMD/issues/204)

**Comment by nigredo-tori** [comment]
So this happened to me again.
Yesterday I had to do a hard shutdown (seems like something was blocking Windows from powering down). This morning I have launched FMD, it didn't show any new chapters and I shut down my PC normally (I do not know if the problem was already present). As I got home after work, I launched FMD only to find that it lost all my favorites again.

1. favorites.ini contains zero records, modification date - this morning, way before powering down.
2. works.ini is empty, modification date - last evening, way before powering down.
3. Account list is intact.
4. Connections tab dropped to default, same with selected websites list. Not sure about other settings.
5. downloadedchapters.db seems intact (an improvement!), but for some reason downloaded chapters don't get highlighted (green) in the GUI.

Seems like this issue should be reopened.
